RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc

```diff
@@ -235,6 +236,109 @@ namespace reco {
+            sigCands_dphi = sigCands_dphi / sigCands_pt;
+         } else {
+            sigCands_dr = sigCands_deta = sigCands_dphi = -0.1;
+         }
+         float isoCands_pt = 0.;
+         float isoCands_dr, isoCands_deta, isoCands_dphi;
+         isoCands_dr = isoCands_deta = isoCands_dphi = 0.;
+         for (auto& j : tau.isolationGammaCands()) {
+            const float dr = reco::deltaR(tau, *j);
+            const float deta = std::abs(tau.eta() - j->eta());
+            const float dphi = std::abs(reco::deltaPhi(tau.phi(), j->phi()));
+            const float pt_ = j->pt();
+            isoCands_dr += dr * pt_;
+            isoCands_deta += deta * pt_;
+            isoCands_dphi += dphi * pt_;
+            isoCands_pt += pt_;
+         }
+         if (isoCands_pt>0.) {
+            isoCands_dr = isoCands_dr / isoCands_pt;
+            isoCands_deta = isoCands_deta / isoCands_pt;
+            isoCands_dphi = isoCands_dphi / isoCands_pt;
+         } else {
+            isoCands_dr = isoCands_deta = isoCands_dphi = -0.1;
+         }
+         mvaInput[9] = isoCands_deta;
+         mvaInput[10] = isoCands_dphi;
+         mvaInput[11] = isoCands_dr;
+         mvaInput[12] = sigCands_deta;
+         mvaInput[13] = sigCands_dphi;
+         mvaInput[14] = sigCands_dr;
+         
+         float e = tau.hcalEnergy()+tau.ecalEnergy();
+         e>0. ? e=tau.ecalEnergy()/e : e=-1.;
+         mvaInput[15] = e;
+         mvaInput[16] = tau.dxy()>=0.?+1:-1;
+         mvaInput[17] = sqrt(abs(tau.dxy()));
+         mvaInput[18] = abs(tau.dxy_Sig());
+         mvaInput[19] = tau.ip3d()>=0.?+1:-1;
+         mvaInput[20] = sqrt(abs(tau.ip3d()));
+         mvaInput[21] = abs(tau.ip3d_Sig());
+         mvaInput[22] = tau.hasSecondaryVertex();
+         mvaInput[23] = decayDistMag; //sqrt(tau.flightLength().Mag2());
+         mvaInput[24] = tau.flightLengthSig();
+         mvaInput[25] = leadingTrackChi2; //tau.leadingTrackNormChi2();
+
+         float thetaGJmax, thetaGJ;
+         thetaGJmax = thetaGJ = 0.;
+         if (decayDistMag>0.) {
+            const float mTau = 1.77686;
+            const float mAOne = tau.p4().M();
+            const float pAOneMag = tau.p();
+            thetaGJmax = (mTau*mTau - mAOne*mAOne) / ( 2. * mTau * pAOneMag);
+            thetaGJmax = asin(thetaGJmax);
+            thetaGJ = (tau.px()*tau.flightLength().x() + tau.py()*tau.flightLength().y() + tau.pz()*tau.flightLength().z()) / (pAOneMag * decayDistMag);
+            thetaGJ = acos(thetaGJ);
+         } else {
+            thetaGJ = -15.;
+            thetaGJmax = -10.;
+         }
+         if (std::isnan(thetaGJ)) thetaGJ = -16.;
+         if (std::isnan(thetaGJmax)) thetaGJmax = -11.;
+         mvaInput[26] = tau.hasSecondaryVertex()?thetaGJ-thetaGJmax:-5.;
+
+         mvaInput[27] = 0;
+         mvaInput[28] = 10.;
+         mvaInput[29] = 10.;
+         if (tau.leadChargedHadrCand().isNonnull()) {
+            if (tau.leadChargedHadrCand()->bestTrack()) {
```

Comment:
For my knowledge, why do you suggest `hasTrackDetails()`?

Reply:
It is to not try to produce a (best)track pointer twice - `hasTrackDetails()` checks conditions allowing to produce track pointer and returns a boolean. But actually it is only cosmetics.

Reply:
Ups, I see only now that you use bare `reco::Candidate` w/o casting to `pat::PackedCandidate` so `bestTrack()` is only available method. In this case it is better to use it as safe casting with `dynamic_cast<>` is rather costfull. Usage of `static_cast<>` can be an alterantive, but can lead to problems in case one uses the producer with taus before slimming, i.e. with `reco::PFCandidate`s. Sorry for the misleading comment.